modules/meshrenderinggl/src/processors/meshrasterizer.cpp

```diff
@@ -1,549 +1,523 @@
+    utilgl::setShaderUniforms(*shader_, *camera_, cameraId_);
+    setLightingUniforms();
+    shader_->setUniform("halfScreenSize", imageSize_ / ivec2(2));
+
+    // update face render settings
+    std::array<TextureUnit, 2> transFuncUnit;
+    for (size_t j = 0; j < faceSettings_.size(); ++j) {
+        const std::string prefix = fmt::format("renderSettings[{}].", j);
+        auto& face = faceSettings_[faceSettings_[1].sameAsFrontFace_.get() ? 0 : j];
+        face.setUniforms(*shader_, prefix);
+
+        const Layer* tfLayer = face.transferFunction_->getData();
+        const LayerGL* transferFunctionGL = tfLayer->getRepresentation<LayerGL>();
+        transferFunctionGL->bindTexture(transFuncUnit[j].getEnum());
+        shader_->setUniform(fmt::format("transferFunction{}", j), transFuncUnit[j].getUnitNumber());
+    }
+
+    // update alpha settings
+    alphaSettings_.setUniforms(*shader_, "alphaSettings.");
+
+    // update other global fragment shader settings
+    shader_->setUniform("silhouetteColor", silhouetteColor_);
+
+    // update geometry shader settings
+    shader_->setUniform("geomSettings.edgeWidth", edgeSettings_.edgeThickness_);
+    shader_->setUniform("geomSettings.triangleNormal",
+                        normalSource_ == NormalSource::GenerateTriangle);
+
+    // call the callback provided by the renderer calling this function
+    setUniformsRenderer(*shader_);
+    {
+        // various OpenGL states: depth, blending, culling
+        utilgl::GlBoolState depthTest(GL_DEPTH_TEST, forceOpaque_);
+        utilgl::DepthMaskState depthMask(forceOpaque_ ? GL_TRUE : GL_FALSE);
+
+        utilgl::CullFaceState culling(
+            !faceSettings_[0].show_ && faceSettings_[1].show_
+                ? GL_FRONT
+                : faceSettings_[0].show_ && !faceSettings_[1].show_ ? GL_BACK : GL_NONE);
+        utilgl::BlendModeState blendModeState(forceOpaque_ ? GL_ONE : GL_SRC_ALPHA,
+                                              forceOpaque_ ? GL_ZERO : GL_ONE_MINUS_SRC_ALPHA);
+
+        // Finally, draw it
+        for (auto mesh : enhancedMeshes_) {
+            MeshDrawerGL::DrawObject drawer{mesh->getRepresentation<MeshGL>(),
+                                            mesh->getDefaultMeshInfo()};
+            utilgl::setShaderUniforms(*shader_, *mesh, "geometry");
+            shader_->setUniform("pickingEnabled", meshutil::hasPickIDBuffer(mesh.get()));
+
+            drawer.draw();
+        }
+    }
+
+    shader_->deactivate();
+}
+
+void MeshRasterizer::MeshRasterization::setLightingUniforms() const {
+    vec3 lightPos;
+    switch (lightSpace_) {
+        case SimpleLightingProperty::Space::VIEW:
+            lightPos = vec3(camera_->getInverseViewMatrix() * vec4(lightPosition_, 1.0f));
+            break;
+        case SimpleLightingProperty::Space::WORLD:
+        default:
+            lightPos = lightPosition_;
+    }
+    shader_->setUniform("lighting.position", lightPos);

```

Comment:
could we not just have the MeshRastersize set all these uniform directly on the shader? why pass them along and then set them, when we could do it directly? we only need to set them when they change or we recompile right?

Reply:
That's a very good point and got rid of a lot of code. Thanks!